### PR TITLE
[wip] pari: add config options

### PR DESCRIPTION
**Pull request by dkwo**
This is such that pari is usable when building sagemath, which requires pthread.

Note: pthread also sets --enable-tls (build the thread-safe version of the library). This tends to slow down the shared library libpari.so by about 25%, so it is suggested to use the static library libpari.a instead. How do I do that?
Arch Linux has `export CFLAGS+=' -flto'`, Alpine has `export CFLAGS="$CFLAGS -flto" # reduce speed losses under pthread`.

### CODE REVIEW

**Comment by ericonr** [comment]
> so it is suggested to use the static library libpari.a instead

I don't understand this part...

Anyway, `CFLAGS=-flto` in the template ought to be enough.

File: srcpkgs/pari/template [hunk 1]
```diff
@@ -19,6 +19,7 @@ checksum=81ecf7d70ccdaae230165cff627c9ce2ec297b8f22f9f40742279d85f86dfcb1
 
 build_options="x11"
 build_options_default="x11"
+CFLAGS=-flto
```
Comment (ericonr) [inline srcpkgs/pari/template:22]: Adding a comment as to why is important, tho :)

**Comment by ericonr** [comment]
@sgn would appreciate a review and ok.

**Comment by sgn** [comment]
Acked.

fwiw, you can take over maintainership.

File: srcpkgs/pari/template (hunk 1 above)
Comment (dkwo) [reply srcpkgs/pari/template:23]: I added a comment in template.

**Comment by dkwo** [comment]
Now `pari` builds and tests fine on `x86_64-musl`:
```
+++ Total bench for gp-sta is 469352
+++ Total bench for gp-dyn is 546837
The following tests were skipped: ellglobalred ellmodulareqn galois galoischartable galpol lfunartin member programming
```
while `giac` builds but fails one test `FAIL: chk_fhan4`.
Can this be fixed again by enlarging `stacksize` @tornaria ?

**Comment by ericonr** [comment]
Increasing stack size should only matter if the test case is segfaulting. Is that the case?

**Comment by dkwo** [comment]
Maybe I'm misinterpreting this?
```
// Using locale /usr/share/locale/
// C
// /usr/share/locale/
// giac
//   ***   Warning: increasing stack size to 1024000.
  ***   the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  *** matdet: Warning: increasing stack size to 2048000.
  ***   at top-level: matdet([-10,-7,6,-7,1,4,-1,-2,-2,-5,1,7,-6,7,-
  ***                 ^----------------------------------------------
  *** matdet: the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  ***   the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  ***   at top-level: matdet([-10,-7,6,-7,1,4,-1,-2,-2,-5,1,7,-6,7,-
  ***                 ^----------------------------------------------
  *** matdet: the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  ***   the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  ***   at top-level: matdet([-10,-7,6,-7,1,4,-1,-2,-2,-5,1,7,-6,7,-
  ***                 ^----------------------------------------------
  *** matdet: the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  ***   the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  ***   at top-level: matdet([-10,-7,6,-7,1,4,-1,-2,-2,-5,1,7,-6,7,-
  ***                 ^----------------------------------------------
  *** matdet: the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  ***   the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  ***   at top-level: matdet([-10,-7,6,-7,1,4,-1,-2,-2,-5,1,7,-6,7,-
  ***                 ^----------------------------------------------
  *** matdet: the thread stack overflows !
  current stack size: 512000 (0.488 Mbytes)
  [hint] set 'threadsizemax' to a nonzero value in your GPRC
  ***   Warning: increasing stack size to 1024000.
  ***   Warning: increasing stack size to 2048000.
  ***   Warning: increasing stack size to 4096000.
  ***   Warning: increasing stack size to 1024000.
  ***   Warning: increasing stack size to 1024000.
  ***   Warning: increasing stack size to 2048000.
49,57d48
< 
< 
< 
< 
< 
< 
< 
< 
< 
61,62c52,53
< undef,
< undef,
---
> 0,
> 0,
73d63
< 
FAIL: chk_fhan4
```

**Comment by ericonr** [comment]
You can probably try increasing it and see what happens ;)

But

>   [hint] set 'threadsizemax' to a nonzero value in your GPRC

seems kinda suspicious? Not sure

**Comment by dkwo** [comment]
I found out that upgrading `giac` results in all checks passed.
I'd still like to add as makedepends `ntl`, which in turn requires `gf2x`.
These are also needed for `flint`.
I will do these separately.

**Comment by tornaria** [comment]
> Maybe I'm misinterpreting this?
> 
> ...
> //   ***   Warning: increasing stack size to 1024000.
>   ***   the thread stack overflows !
>   current stack size: 512000 (0.488 Mbytes)
>   [hint] set 'threadsizemax' to a nonzero value in your GPRC
>   *** matdet: Warning: increasing stack size to 2048000.
>   ***   at top-level: matdet([-10,-7,6,-7,1,4,-1,-2,-2,-5,1,7,-6,7,-
>   ***                 ^----------------------------------------------
>   *** matdet: the thread stack overflows !
>   current stack size: 512000 (0.488 Mbytes)
>   [hint] set 'threadsizemax' to a nonzero value in your GPRC
>   ***   the thread stack overflows !
> ...

I think you are confusing the actual stack used by the cpu, with the pari stack where computations are done (this is the basic design of pari using a stack for allocations instead of GC, etc).

Usually (with single threaded pari) when the stack overflows pari doubles the stack and retries the computation. I think this already happened for this test. However it seems that in multi-threaded pari the doubling-the-stack behaviour is limited by `threadsizemax` (I guess to avoid using up a lot of memory), so the stack size, originally at 512K, is increased to 1024K, but `threadsizemax` prevents that (note that the second time the current stack size is still 512K).

I will have a look into this. It might be: (a) that the particular computation giac is doing in this test is hard and it really require more than 512k of stack (b) there's a "bug" in pari, maybe some routine in pari is not checking the stack to reallocate often enough so it ends up not working in limited stack space.

Regarding pthread pari: I'd like to discuss a bit more what's the right way to proceed. Indeed for lots of applications having a pthread pari is a good thing (I use a 36-core machine for my computations so I'd definitely take advantage of this). However, I've noticed a slowdown by a factor of 2 with pthread pari in some of my computations, so I often use single-threaded pari and arrange my computation so it can be done by 36 independent pari processes. I think the reason is that TLS is _slow_.

I wonder if it would make sense to offer different versions of pari (single thread / multi thread). At the very least, I'd make pthread a build option so that one can easily compile one or the other as needed.

More generally: in math software is often important to have different versions of the same software available. I don't know if there is a reasonable way to offer this in void. Also might be desirable: offering a way to compile math software with -march=native so it can take advantage of avx, etc.

Is there a way void could support something like "subarchitectures"? Say `x86_64-avx` where some packages are compiled for this architecture and the rest default to the packages for `x86_64`.

**Comment by tornaria** [comment]
> This is such that pari is usable when building sagemath, which requires pthread.
As a matter of fact, I don't think sagemath requires pari built with pthread, or does it? (maybe it does need TLS at least? I don't know)

Anyway, sagemath won't work with pari-2.13 (for now, see https://trac.sagemath.org/ticket/30801) but that's for a different reason.

I run check as a "benchmark" to have hard data, but so far it seems to support what I thought:
 - using `pthreads` has a 25-45 % performance hit (25% for `gp-sta`, 45% for `gp-dyn`)
 - using `CFLAGS=-flto` doesn't change anything (even make `gp-dyn` slower).

A very stupid (non-)benchmark is to run
```
$ echo "sum(i=1,100 000 000, i);" | time gp -q
```
This doesn't substitute a proper benchmark, but at least it takes less than 5s to show a significant slowdown: it takes ~ 4.25s for me with the current binary from `pari-2.13.1_1`, but ~ 8.5s with `--mt=pthread` and more than 9s with `-flto` (less if using gp-sta, but we are shipping gp-dyn).

Note that all this "benchmark" is using a single-thread, because most (all?) of pari code which is written in C is single-thread. If you actually USE multiple threads you'll pay another 50% penalty. For instance:
```
echo "parsum(j=0,99,sum(i=j*1 000 000 + 1, (j+1)*1 000 000, i));" | time gp -q
```
Takes:
 - with single-thread gp-dyn: still 4.25s (no different than above)
 - with single-thread gp-sta: ~ 4s
 - with pthread gp-dyn: 12.87s user = 2.40s wall (8 core)
 - with pthread gp-sta: 7.28s user = 1.37s wall (8 core)

I'm not willing to pay a 3x penalty for using multi-thread. I can parallelize my programs in a different way, but of course not all tasks are suitable for that.

Here's a proposal (RFC):
 - start building and shipping gp-sta instead of gp-dyn (it's not clear what's the advantage of gp-dyn other than disk space).
 - add a build option to pari so that it's easy to compile pari with pthread support. It's even more important that the pthread version compiles gp-sta as the penalty is higher here.
 - start shipping the static library so external programs can static link with pari.
 - consider actually compiling both single-thread and multi-thread pari. The libraries have different SONAMES so that would not be a problem. As for the gp binary, we could ship them in separate packages, either conflicting or with different names and alternatives.

What do you think?

I'm willing to do all of the above and take over the package if the plan above sounds reasonable.

**Comment by tornaria** [comment]
> I found out that upgrading `giac` results in all checks passed.
> I'd still like to add as makedepends `ntl`, which in turn requires `gf2x`.
> These are also needed for `flint`.
> I will do these separately.

I have templates for givaro, fflas-ffpack, linbox, gf2x, ntl, flintlib, arb, mpfi, eclib, flintqs, lcalc. They are a bit old but were working ok to compile sagemath 9.1. I'll try to update and make PR for all of that. I'd really like to have a working sagemath in void.

I also have a WIP template for gap, but it seemed more difficult so it's not really working at this time.

OTOH, note that using system packages to build sage is a bit "dangerous", because some system updates will break your build of sage. This is the main reason why it's important to have sagemath in void: so that relevant updates will revbump and recompile it. Otherwise it's safer to use the minimal set of system packages.

**Comment by ericonr** [comment]
Two quick nits on the current PR: the `pari` commit messages should be squashed together, and the `giac` commit should include `update to $version`, instead of just `update` :)

@tornaria, thanks for the thorough analysis.

> I think the reason is that TLS is slow.

This shouldn't affect things so much, unless they made the terrible decision of using `__thread` globals for everything? Or worse, normal global with `pthread_mutex` calls around accesses...

> I wonder if it would make sense to offer different versions of pari (single thread / multi thread). At the very least, I'd make pthread a build option so that one can easily compile one or the other as needed.

That sounds reasonable, especially since there seems to be such a difference.

> Is there a way void could support something like "subarchitectures"? Say x86_64-avx where some packages are compiled for this architecture and the rest default to the packages for x86_64.

Maybe as a build option, at most. I don't believe we wish to carry such packages, because it can increase the maintenance burden a lot. If the threaded version becomes necessary for sage (as a system package) but the performance loss is too high, we can consider it then, but I'm against it when it comes to architecture. Ideally, they would compile opimitized versions of some files and choose between them at runtime...

**Comment by ericonr** [comment]
> OTOH, note that using system packages to build sage is a bit "dangerous", because some system updates will break your build of sage. This is the main reason why it's important to have sagemath in void: so that relevant updates will revbump and recompile it. Otherwise it's safer to use the minimal set of system packages.

If you have a local sage build installed via XBPS, it should catch shlib errors, unless some dependencies break ABI/compatibility without soname bumps?

Anyway, tagging in @ahesford because it might be near his ballpark.

**Comment by dkwo** [comment]
@tornaria Thanks, I'll be fine doing as you propose. Also, adding those packages would be great.
@ericonr I made changes, but feel free to close this if we decide to do otherwise.
My reason was that in sage-9.3.beta9 I get
```
cat build/pkgs/pari/spkg-configure.m4 | grep pthread
                     AC_MSG_CHECKING([whether pari is configured with pthreads])
                        [[return strcmp(PARI_MT_ENGINE, "pthread") != 0]])],
```

**Comment by tornaria** [comment]
> @tornaria Thanks, I'll be fine doing as you propose. Also, adding those packages would be great.
> @ericonr I made changes, but feel free to close this if we decide to do otherwise.
> My reason was that in sage-9.3.beta9 I get
> 
> ```
> cat build/pkgs/pari/spkg-configure.m4 | grep pthread
>                      AC_MSG_CHECKING([whether pari is configured with pthreads])
>                         [[return strcmp(PARI_MT_ENGINE, "pthread") != 0]])],
> ```
Thanks for pointing this out, I had missed it. This was introduced in sage-9.3-beta9 and there is no reason given. I created https://trac.sagemath.org/ticket/31572 about it.

For the moment we can't use system pari because 2.13 is not supported in sagemath (yet), so I would say leave it single thread by default for now. We can figure out how to proceed once the dust settles on that sagemath ticket and pari-2.13 gets supported (it seems not for sage-9.3, unfortunately).

**Comment by dkwo** [comment]
> We can figure out how to proceed once the dust settles on that sagemath ticket and pari-2.13 gets supported (it seems not for sage-9.3, unfortunately).

Sounds good.
I changed title to wip, but feel free to close it.

**Comment by dkwo** [comment]
Note to self: new math packages (lcalc, eclib) will need revbump; checkdepends for pari.

**Comment by tornaria** [comment]
I think this is superseded by #32453 (already merged).

**Comment by ericonr** [comment]
@dkwo are you still interested in the `giac` fixes?

**Comment by tornaria** [comment]
> @dkwo are you still interested in the `giac` fixes?

See #32558 which contains the giac update.

**Comment by dkwo** [comment]
Superseded by the two mentioned PRs, closing.

**Pull request closed without merging**